File: internal/regime/service.go
package regime

import (
	"context"
	"encoding/json"
	"fmt"
	"log/slog"
	"strconv"
	"sync"
	"time"

	"github.com/nats-io/nats.go/jetstream"

	"github.com/noeljackson/stocks/internal/domain"
	"github.com/noeljackson/stocks/internal/platform/bus"
	"github.com/noeljackson/stocks/internal/platform/store"
	"github.com/noeljackson/stocks/internal/platform/subjects"
)

// macroObservation is the schema produced by the FRED adapter.
type macroObservation struct {
	Series string `json:"series"`
	Date   string `json:"date"`
	Value  string `json:"value"`
}

// Service maintains the latest indicator snapshot and recomputes regime on
// every macro update. It is deterministic and stateless across restarts —
// the in-memory snapshot is rebuilt from the most recent ingest_event row per
// series at boot (if no events yet, it starts empty; classify → neutral).
type Service struct {
	DB  *store.DB
	Bus *bus.Bus
	Log *slog.Logger

	mu   sync.Mutex
	snap map[string]float64
	last domain.Regime
	lastCap bool
}

func New(db *store.DB, b *bus.Bus, log *slog.Logger) *Service {
	return &Service{DB: db, Bus: b, Log: log, snap: map[string]float64{}, last: domain.RegimeNeutral}
}

// Run wires the regime classifier:
//   - ensures the MARKET stream (publish target for regime.*) exists,
//   - warm-starts the in-memory snapshot from ingest_event,
//   - binds a durable JetStream consumer on INGEST/ingest.macro,
//   - blocks until ctx is cancelled.
func (s *Service) Run(ctx context.Context) error {
	if err := s.Bus.EnsureStream(ctx, subjects.StreamMarket, "regime.*", "discovery.*"); err != nil {
		return fmt.Errorf("ensure MARKET: %w", err)
	}
	if err := s.warmStart(ctx); err != nil {
		s.Log.Warn("warm-start failed; starting empty", "err", err)
	}
	stop, err := s.Bus.Consume(ctx, subjects.StreamIngest, "regime-classifier", subjects.IngestMacro,
		func(m jetstream.Msg) error { return s.onMacro(ctx, m.Data()) })
	if err != nil {
		return err
	}
	defer stop()
	s.Log.Info("regime classifier consuming", "stream", subjects.StreamIngest, "filter", subjects.IngestMacro)
	<-ctx.Done()
	return nil
}

// observation is the parsed-and-mapped form of a macro message.
type observation struct {
	Series, Name string
	Value        float64
	Date         string
}

// parseObservation decodes the FRED adapter payload and applies the
// series-name → indicator-name mapping. Returns ok=false for malformed
// (poison) messages; the caller should ack-and-drop those.
func parseObservation(data []byte) (observation, bool) {
	var raw macroObservation
	if err := json.Unmarshal(data, &raw); err != nil {
		return observation{}, false
	}
	if raw.Series == "" {
		return observation{}, false
	}
	v, err := strconv.ParseFloat(raw.Value, 64)
	if err != nil {
		return observation{}, false
	}
	name, ok := FREDSeriesToIndicator[raw.Series]
	if !ok {
		name = raw.Series
	}
	return observation{Series: raw.Series, Name: name, Value: v, Date: raw.Date}, true
}

// applyObservation updates the in-memory snapshot with obs and returns the
// classification + a snapshot copy + whether the regime/capitulation flipped.
// Pure (besides the locked snapshot mutation) — no DB, no NATS, no clock.
func (s *Service) applyObservation(cfg Config, obs observation) (Result, map[string]float64, bool) {
	s.mu.Lock()
	s.snap[obs.Name] = obs.Value
	snap := make(map[string]float64, len(s.snap))
	for k, v := range s.snap {
		snap[k] = v
	}
	prevReg, prevCap := s.last, s.lastCap
	s.mu.Unlock()

	res := Classify(cfg, snap)

	s.mu.Lock()
	s.last = res.Regime
	s.lastCap = res.Capitulation
	s.mu.Unlock()

	return res, snap, res.Regime != prevReg || res.Capitulation != prevCap
}

func (s *Service) onMacro(ctx context.Context, data []byte) error {
	obs, ok := parseObservation(data)
	if !ok {
		s.Log.Warn("dropping malformed macro message")
		return nil // poison: ack
	}
	cfg, ver, err := s.loadConfig(ctx)
	if err != nil {
		s.Log.Error("load regime config", "err", err)
		return err // transient: let JetStream redeliver
	}
	res, snap, changed := s.applyObservation(cfg, obs)

	asOf := time.Now().UTC()
	ind, _ := json.Marshal(snap)
	if err := s.DB.UpsertMarketState(ctx, asOf, string(res.Regime), res.Capitulation, ind, ver); err != nil {
		s.Log.Error("persist market_state", "err", err)
		return err
	}
	out, _ := json.Marshal(map[string]any{
		"as_of":          asOf,
		"regime":         res.Regime,
		"capitulation":   res.Capitulation,
		"matched":        res.Matched,
		"config_version": ver,
		"trigger":        map[string]any{"series": obs.Series, "name": obs.Name, "value": obs.Value, "date": obs.Date},
	})
	if err := s.Bus.Publish(ctx, subjects.RegimeState, out); err != nil {
		s.Log.Error("publish regime.state", "err", err)
		return err
	}
	if changed {
		s.Log.Info("regime change", "regime", res.Regime, "capitulation", res.Capitulation)
		if res.Capitulation {
			_ = s.Bus.Publish(ctx, subjects.RegimeCapitulation, out)
		}
	}
	return nil
}

// warmStart rebuilds the snapshot from the latest FRED observation per series
// already in ingest_event (so a restart doesn't lose state until the next poll).
func (s *Service) warmStart(ctx context.Context) error {
	rows, err := s.DB.Pool.Query(ctx, `
		SELECT DISTINCT ON (payload->>'series') payload
		  FROM ingest_event
		 WHERE source='fred'
		 ORDER BY payload->>'series', ingested_at DESC`)
	if err != nil {
		return err
	}
	defer rows.Close()
	for rows.Next() {
		var raw []byte
		if err := rows.Scan(&raw); err != nil {
			return err
		}
		var obs macroObservation
		if err := json.Unmarshal(raw, &obs); err != nil {
			continue
		}
		v, err := strconv.ParseFloat(obs.Value, 64)
		if err != nil {
			continue
		}
		name, ok := FREDSeriesToIndicator[obs.Series]
		if !ok {
			name = obs.Series
		}
		s.snap[name] = v
	}
	return rows.Err()
}

func (s *Service) loadConfig(ctx context.Context) (Config, int, error) {
	body, ver, err := s.DB.ActiveConfig(ctx, "regime")
	if err != nil {
		return Config{}, 0, err
	}
	var cfg Config
	if err := json.Unmarshal(body, &cfg); err != nil {
		return Config{}, 0, err
	}
	return cfg, ver, nil
}

File: src/regime/service.rs
//! Regime classifier service: subscribes to ingest.macro via a durable
//! consumer, maintains an in-memory snapshot, persists market_state on every
//! update, publishes regime.state.

use std::collections::HashMap;
use std::sync::Arc;

use anyhow::Result;
use chrono::{DateTime, Utc};
use serde::Deserialize;
use sqlx::Row;
use tokio::sync::Mutex;
use tracing::{error, info, warn};

use super::{Config, Outcome, classify, fred_series_to_indicator};
use crate::platform::bus::{Bus, ConsumerHandle};
use crate::platform::domain::Regime;
use crate::platform::store::Store;
use crate::platform::subjects;

#[derive(Deserialize)]
struct MacroObs {
    series: String,
    date: String,
    value: String,
}

#[derive(Default)]
struct State {
    snap: HashMap<String, f64>,
    last_regime: Option<Regime>,
    last_capitulation: bool,
}

/// Run the regime classifier service.
pub async fn run(store: Store, bus: Bus) -> Result<ConsumerHandle> {
    bus.ensure_stream(subjects::STREAM_MARKET, &["regime.*", "discovery.*"])
        .await?;

    let state = Arc::new(Mutex::new(State::default()));
    warm_start(&store, &state).await.unwrap_or_else(|e| {
        warn!(error = %e, "warm-start failed; starting empty");
    });

    let store = Arc::new(store);
    let bus = Arc::new(bus);
    let bus_consume = bus.clone();

    let handle = bus_consume
        .consume(
            subjects::STREAM_INGEST,
            "regime-classifier",
            subjects::INGEST_MACRO,
            {
                let store = store.clone();
                let bus = bus.clone();
                let state = state.clone();
                move |msg| {
                    let store = store.clone();
                    let bus = bus.clone();
                    let state = state.clone();
                    async move { on_macro(&store, &bus, &state, &msg.payload).await }
                }
            },
        )
        .await?;
    info!(
        stream = subjects::STREAM_INGEST,
        filter = subjects::INGEST_MACRO,
        "regime classifier consuming"
    );
    Ok(handle)
}

async fn on_macro(store: &Store, bus: &Bus, state: &Mutex<State>, data: &[u8]) -> Result<()> {
    let Ok(obs) = serde_json::from_slice::<MacroObs>(data) else {
        warn!("dropping malformed macro message");
        return Ok(());
    };
    if obs.series.is_empty() {
        warn!("macro missing series; dropping");
        return Ok(());
    }
    let Ok(val) = obs.value.parse::<f64>() else {
        warn!(series = %obs.series, value = %obs.value, "non-numeric macro value; dropping");
        return Ok(());
    };
    let name = fred_series_to_indicator(&obs.series).to_string();

    // Load config every message (cheap; lets config bumps take effect immediately).
    let (cfg_json, cfg_ver) = store.active_config("regime").await?;
    let cfg: Config = serde_json::from_value(cfg_json)?;

    let outcome = {
        let mut s = state.lock().await;
        s.snap.insert(name.clone(), val);
        let snap = s.snap.clone();
        let prev_regime = s.last_regime;
        let prev_cap = s.last_capitulation;
        let outcome: Outcome = classify(&cfg, &snap);
        s.last_regime = Some(outcome.regime);
        s.last_capitulation = outcome.capitulation;
        OutcomeWithChange {
            outcome,
            prev_regime,
            prev_capitulation: prev_cap,
            changed: prev_regime != Some(s.last_regime.unwrap()) || prev_cap != s.last_capitulation,
        }
    };

    let as_of = Utc::now();
    let indicators_json = serde_json::to_value(&outcome.outcome.indicators)?;
    let matched_json = serde_json::to_value(&outcome.outcome.matched)?;
    let trigger_json = serde_json::json!({
        "series": obs.series,
        "name": name,
        "value": val,
        "date": obs.date,
    });
    if let Err(e) = store
        .upsert_market_state(
            as_of,
            outcome.outcome.regime.as_str(),
            outcome.outcome.capitulation,
            &indicators_json,
            cfg_ver,
        )
        .await
    {
        error!(error = %e, "persist market_state");
        return Err(e);
    }
    if outcome.changed {
        upsert_regime_evidence_item(
            store,
            as_of,
            outcome.prev_regime,
            outcome.outcome.regime,
            outcome.prev_capitulation,
            outcome.outcome.capitulation,
            &trigger_json,
            &matched_json,
            cfg_ver,
        )
        .await?;
    }
    let publish_body = serde_json::json!({
        "as_of": as_of,
        "regime": outcome.outcome.regime,
        "capitulation": outcome.outcome.capitulation,
        "matched": outcome.outcome.matched,
        "config_version": cfg_ver,
        "trigger": trigger_json,
    });
    bus.publish(subjects::REGIME_STATE, publish_body.to_string().as_bytes())
        .await?;
    if outcome.changed {
        info!(
            regime = %outcome.outcome.regime.as_str(),
            capitulation = outcome.outcome.capitulation,
            "regime change"
        );
        if outcome.outcome.capitulation {
            bus.publish(
                subjects::REGIME_CAPITULATION,
                publish_body.to_string().as_bytes(),
            )
            .await?;
        }
    }
    Ok(())
}

struct OutcomeWithChange {
    outcome: Outcome,
    prev_regime: Option<Regime>,
    prev_capitulation: bool,
    changed: bool,
}

pub(super) fn regime_evidence_summary(
    previous: Option<Regime>,
    current: Regime,
    previous_capitulation: bool,
    current_capitulation: bool,
    trigger_name: &str,
    trigger_value: f64,
) -> String {
    let trigger = format!("{trigger_name}={trigger_value:.2}");
    match previous {
        None => format!(
            "Market regime initialized {} after {trigger}",
            current.as_str()
        ),
        Some(prev) if prev != current => format!(
            "Market regime changed {} -> {} after {trigger}",
            prev.as_str(),
            current.as_str()
        ),
        Some(_) if previous_capitulation != current_capitulation => format!(
            "Market regime {} capitulation changed {} -> {} after {trigger}",
            current.as_str(),
            previous_capitulation,
            current_capitulation
        ),
        Some(_) => format!(
            "Market regime {} refreshed after {trigger}",
            current.as_str()
        ),
    }
}

pub(super) fn regime_evidence_strength(capitulation: bool) -> f64 {
    if capitulation { 0.9 } else { 0.7 }
}

pub(super) fn regime_evidence_polarity(regime: Regime) -> f64 {
    match regime {
        Regime::RiskOn => 0.5,
        Regime::Neutral => 0.0,
        Regime::RiskOff => -0.5,
    }
}

async fn upsert_regime_evidence_item(
    store: &Store,
    as_of: DateTime<Utc>,
    prev_regime: Option<Regime>,
    regime: Regime,
    prev_capitulation: bool,
    capitulation: bool,
    trigger: &serde_json::Value,
    matched: &serde_json::Value,
    config_version: i32,
) -> Result<()> {
    let trigger_name = trigger
        .get("name")
        .and_then(serde_json::Value::as_str)
        .unwrap_or("macro");
    let trigger_value = trigger
        .get("value")
        .and_then(serde_json::Value::as_f64)
        .unwrap_or(0.0);
    let summary = regime_evidence_summary(
        prev_regime,
        regime,
        prev_capitulation,
        capitulation,
        trigger_name,
        trigger_value,
    );
    let source_ref = serde_json::json!({
        "table": "market_state",
        "as_of": as_of,
        "previous_regime": prev_regime.map(|r| r.as_str()),
        "previous_capitulation": prev_capitulation,
        "regime": regime.as_str(),
        "capitulation": capitulation,
        "trigger": trigger,
        "matched": matched,
        "config_version": config_version,
    });
    sqlx::query(
        r#"INSERT INTO evidence_item
             (symbol, kind, observed_at, source, source_id, source_ref,
              summary, strength, polarity)
           VALUES (
             'MARKET', 'regime', $1, 'regime_classifier', $2, $3,
             $4, $5, $6
           )
           ON CONFLICT (source, source_id) DO UPDATE SET
             source_ref = evidence_item.source_ref || EXCLUDED.source_ref,
             summary = EXCLUDED.summary,
             strength = EXCLUDED.strength,
             polarity = EXCLUDED.polarity,
             updated_at = now()"#,
    )
    .bind(as_of)
    .bind(format!("market_state:{}", as_of.to_rfc3339()))
    .bind(source_ref)
    .bind(summary)
    .bind(regime_evidence_strength(capitulation))
    .bind(regime_evidence_polarity(regime))
    .execute(&store.pool)
    .await?;
    Ok(())
}

/// Rebuilds the snapshot from the latest FRED observation per series already
/// in `ingest_event` so a restart doesn't lose state until the next poll.
async fn warm_start(store: &Store, state: &Mutex<State>) -> Result<()> {
    let rows = sqlx::query(
        r#"SELECT DISTINCT ON (payload->>'series') payload
             FROM ingest_event
            WHERE source = 'fred'
         ORDER BY payload->>'series', ingested_at DESC"#,
    )
    .fetch_all(&store.pool)
    .await?;
    let mut s = state.lock().await;
    for row in rows {
        let payload: serde_json::Value = row.try_get("payload")?;
        let Ok(obs) = serde_json::from_value::<MacroObs>(payload) else {
            continue;
        };
        let Ok(v) = obs.value.parse::<f64>() else {
            continue;
        };
        let name = fred_series_to_indicator(&obs.series).to_string();
        s.snap.insert(name, v);
    }
    Ok(())
}
